### ISSUE TITLE
Include :: 'Index was outside the bounds of the array' thrown for some cases of multi include queries

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFramework#1488]
Repro:

```
    public class Level1
    {
        public int Id1 { get; set; }
        public string Name { get; set; }

        public Level2 OneToOne_Required_PK { get; set; }
        public Level2 OneToOne_Optional_PK { get; set; }

        public Level2 OneToOne_Required_FK { get; set; }
        public Level2 OneToOne_Optional_FK { get; set; }

        public ICollection<Level2> OneToMany_Optional { get; set; }
    }

    public class Level2
    {
        public int Id2 { get; set; }
        public string Name { get; set; }

        public int Level1_Required_Id { get; set; }
        public int? Level1_Optional_Id { get; set; }

        public Level3 OneToOne_Required_PK { get; set; }
        public Level3 OneToOne_Optional_PK { get; set; }

        public Level3 OneToOne_Required_FK { get; set; }
        public Level3 OneToOne_Optional_FK { get; set; }

        public ICollection<Level3> OneToMany_Optional { get; set; }

        public Level1 OneToOne_Required_PK_Inverse { get; set; }
        public Level1 OneToOne_Optional_PK_Inverse { get; set; }
        public Level1 OneToOne_Required_FK_Inverse { get; set; }
        public Level1 OneToOne_Optional_FK_Inverse { get; set; }

        public Level1 OneToMany_Required_Inverse { get; set; }
        public Level1 OneToMany_Optional_Inverse { get; set; }
    }

    public class Level3
    {
        public int Id3 { get; set; }
        public string Name { get; set; }

        public int Level2_Required_Id { get; set; }
        public int? Level2_Optional_Id { get; set; }

        public Level2 OneToOne_Required_FK_Inverse { get; set; }
    }

    public class MyContext : DbContext
    {
        public DbSet<Level1> LevelOne { get; set; }
        public DbSet<Level2> LevelTwo { get; set; }
        public DbSet<Level3> LevelThree { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            modelBuilder.Entity<Level1>().Key(e => e.Id1);
            modelBuilder.Entity<Level2>().Key(e => e.Id2);
            modelBuilder.Entity<Level3>().Key(e => e.Id3);

            modelBuilder.Entity<Level1>().Property(e => e.Id1).GenerateValueOnAdd(false);
            modelBuilder.Entity<Level2>().Property(e => e.Id2).GenerateValueOnAdd(false);

            modelBuilder.Entity<Level1>().HasOne(e => e.OneToOne_Required_PK).WithOne(e => e.OneToOne_Required_PK_Inverse).ReferencedKey<Level1>(e => e.Id1).Required(true);
            modelBuilder.Entity<Level1>().HasOne(e => e.OneToOne_Optional_PK).WithOne(e => e.OneToOne_Optional_PK_Inverse).ReferencedKey<Level1>(e => e.Id1).Required(false);

            modelBuilder.Entity<Level1>().HasOne(e => e.OneToOne_Required_FK).WithOne(e => e.OneToOne_Required_FK_Inverse).ForeignKey<Level2>(e => e.Level1_Required_Id).Required(true);
            modelBuilder.Entity<Level1>().HasOne(e => e.OneToOne_Optional_FK).WithOne(e => e.OneToOne_Optional_FK_Inverse).ForeignKey<Level2>(e => e.Level1_Optional_Id).Required(false);

            modelBuilder.Entity<Level1>().HasMany(e => e.OneToMany_Optional).WithOne(e => e.OneToMany_Optional_Inverse).Required(false);

            modelBuilder.Entity<Level2>().Property(e => e.Id2).GenerateValueOnAdd(false);
            modelBuilder.Entity<Level2>().Property(e => e.Id2).GenerateValueOnAdd(false);

            modelBuilder.Entity<Level2>().HasOne(e => e.OneToOne_Required_FK).WithOne(e => e.OneToOne_Required_FK_Inverse).ForeignKey<Level3>(e => e.Level2_Required_Id).Required(true);

            modelBuilder.Entity<Level2>().HasOne(e => e.OneToOne_Required_FK).WithOne(e => e.OneToOne_Required_FK_Inverse).ForeignKey<Level3>(e => e.Level2_Required_Id).Required(true);

            modelBuilder.Entity<Level3>().Property(e => e.Id3).GenerateValueOnAdd(false);
            modelBuilder.Entity<Level3>().Property(e => e.Id3).GenerateValueOnAdd(false);
        }

        protected override void OnConfiguring(DbContextOptions options)
        {
            options.UseSqlServer(new SqlConnectionStringBuilder { DataSource = ".", InitialCatalog = "Repro1488", MultipleActiveResultSets = true, IntegratedSecurity = true }.ToString());
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new MyContext())
            {
                ctx.Database.EnsureDeleted();
                ctx.Database.EnsureCreated();

                var l1_01 = new Level1 { Id1 = 1, Name = "L1 01", };
                var l1_02 = new Level1 { Id1 = 2, Name = "L1 02", };
                var l1_03 = new Level1 { Id1 = 3, Name = "L1 03", };
                var l1_04 = new Level1 { Id1 = 4, Name = "L1 04", };

                var l2_01 = new Level2 { Id2 = 1, Name = "L2 01", };
                var l2_02 = new Level2 { Id2 = 2, Name = "L2 02", };
                var l2_03 = new Level2 { Id2 = 3, Name = "L2 03", };
                var l2_04 = new Level2 { Id2 = 4, Name = "L2 04", };

                var l3_01 = new Level3 { Id3 = 1, Name = "L3 01", };
                var l3_02 = new Level3 { Id3 = 2, Name = "L3 02", };
                var l3_03 = new Level3 { Id3 = 3, Name = "L3 03", };
                var l3_04 = new Level3 { Id3 = 4, Name = "L3 04", };

                var l1s = new[] { l1_01, l1_02, l1_03, l1_04, }.ToArray();
                var l2s = new[] { l2_01, l2_02, l2_03, l2_04, }.ToArray();
                var l3s = new[] { l3_01, l3_02, l3_03, l3_04, }.ToArray();

                ctx.LevelOne.Add(l1s);
                ctx.LevelTwo.Add(l2s);
                ctx.LevelThree.Add(l3s);

                l1s[0].OneToOne_Required_PK = l2s[0];
                l1s[1].OneToOne_Required_PK = l2s[1];
                l1s[2].OneToOne_Required_PK = l2s[2];
                l1s[3].OneToOne_Required_PK = l2s[3];

                l1s[0].OneToOne_Required_FK = l2s[3];
                l1s[1].OneToOne_Required_FK = l2s[2];
                l1s[2].OneToOne_Required_FK = l2s[1];
                l1s[3].OneToOne_Required_FK = l2s[0];

                l2s[0].OneToOne_Required_PK = l3s[0];
                l2s[1].OneToOne_Required_PK = l3s[1];
                l2s[2].OneToOne_Required_PK = l3s[2];
                l2s[3].OneToOne_Required_PK = l3s[3];

                l2s[0].OneToOne_Required_FK = l3s[3];
                l2s[1].OneToOne_Required_FK = l3s[2];
                l2s[2].OneToOne_Required_FK = l3s[1];
                l2s[3].OneToOne_Required_FK = l3s[0];

                ctx.SaveChanges();

                l1s[0].OneToOne_Optional_PK = l2s[0];
                l1s[2].OneToOne_Optional_PK = l2s[2];

                l1s[1].OneToOne_Optional_FK = l2s[2];
                l1s[3].OneToOne_Optional_FK = l2s[0];

                l1s[0].OneToMany_Optional = new List<Level2> { l2s[1], l2s[3], };

                l2s[0].OneToOne_Optional_PK = l3s[0];
                l2s[2].OneToOne_Optional_PK = l3s[2];

                l2s[1].OneToOne_Optional_FK = l3s[2];
                l2s[3].OneToOne_Optional_FK = l3s[0];

                l2s[0].OneToMany_Optional = new List<Level3> { l3s[1], l3s[3], };

                ctx.SaveChanges();
            }

            using (var ctx = new MyContext())
            {
                ctx.LevelOne
                    .Include(e => e.OneToMany_Optional)
                    .ThenInclude(e => e.OneToMany_Optional_Inverse)
                    .ThenInclude(e => e.OneToOne_Optional_PK)
                    .ThenInclude(e => e.OneToOne_Required_FK).ToList();

                ctx.LevelOne
                    .Include(e => e.OneToMany_Optional)
                    .ThenInclude(e => e.OneToMany_Optional_Inverse)
                    .ThenInclude(e => e.OneToOne_Optional_PK)
                    .ThenInclude(e => e.OneToOne_Required_FK_Inverse).ToList();
            }
        }
    }
```

throws:

Unhandled Exception: System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Microsoft.Data.Entity.Metadata.ObjectArrayValueReader.IsNull(Int32 index)
in D:\k\EntityFramework\src\EntityFramework.Core\Metadata\ObjectArrayValueReader
.cs:line 22
   at Microsoft.Data.Entity.Relational.Query.OffsetValueReaderDecorator.IsNull(I
nt32 index) in D:\k\EntityFramework\src\EntityFramework.Relational\Query\OffsetV
alueReaderDecorator.cs:line 25
   at Microsoft.Data.Entity.ChangeTracking.SimpleEntityKeyFactory`1.Create(IEnti
tyType entityType, IReadOnlyList`1 properties, IValueReader valueReader) in D:\k
\EntityFramework\src\EntityFramework.Core\ChangeTracking\SimpleEntityKeyFactory.
cs:line 18
   at Microsoft.Data.Entity.Query.QueryBuffer.IncludeCore(Object entity, INaviga
tion navigation, EntityKey& primaryKey, List`1& bufferedEntities, Func`2& relate
dKeyFactory) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.
cs:line 294
   at Microsoft.Data.Entity.Query.QueryBuffer.Include(Object entity, IReadOnlyLi
st`1 navigationPath, IReadOnlyList`1 relatedValueReaders, Int32 currentNavigatio
nIndex) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:li
ne 176
   at Microsoft.Data.Entity.Query.QueryBuffer.<>c__DisplayClass14_0.<Include>b__
0(IValueReader valueReader) in D:\k\EntityFramework\src\EntityFramework.Core\Que
ry\QueryBuffer.cs:line 193
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Microsoft.Data.Entity.Query.QueryBuffer.Include(Object entity, IReadOnlyLi
st`1 navigationPath, IReadOnlyList`1 relatedValueReaders, Int32 currentNavigatio
nIndex) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:li
ne 184
   at Microsoft.Data.Entity.Query.QueryBuffer.<>c__DisplayClass14_0.<Include>b__
0(IValueReader valueReader) in D:\k\EntityFramework\src\EntityFramework.Core\Que
ry\QueryBuffer.cs:line 193
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Microsoft.Data.Entity.Query.QueryBuffer.Include(Object entity, IReadOnlyLi
st`1 navigationPath, IReadOnlyList`1 relatedValueReaders, Int32 currentNavigatio
nIndex) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:li
ne 184
   at Microsoft.Data.Entity.Query.QueryBuffer.<>c__DisplayClass14_0.<Include>b__
0(IValueReader valueReader) in D:\k\EntityFramework\src\EntityFramework.Core\Que
ry\QueryBuffer.cs:line 193
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Microsoft.Data.Entity.Query.QueryBuffer.Include(Object entity, IReadOnlyLi
st`1 navigationPath, IReadOnlyList`1 relatedValueReaders, Int32 currentNavigatio
nIndex) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:li
ne 184
   at Microsoft.Data.Entity.Query.QueryBuffer.Include(Object entity, IReadOnlyLi
st`1 navigationPath, IReadOnlyList`1 relatedValueReaders) in D:\k\EntityFramewor
k\src\EntityFramework.Core\Query\QueryBuffer.cs:line 154
   at Microsoft.Data.Entity.Relational.Query.QueryMethodProvider.<>c__DisplayCla
ss11_0`1.<_Include>b__2(T qss) in D:\k\EntityFramework\src\EntityFramework.Relat
ional\Query\QueryMethodProvider.cs:line 92
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at Microsoft.Data.Entity.Relational.Utilities.EnumerableExtensions.<Finally>d
__0`1.MoveNext() in D:\k\EntityFramework\src\EntityFramework.Relational\Utilitie
s\EnumerableExtensions.cs:line 15
   at System.Linq.Enumerable.<SelectManyIterator>d__8`2.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at Microsoft.Data.Entity.Query.EntityQueryExecutor.EnumerableExceptionInterce
ptor`1.EnumeratorExceptionInterceptor.MoveNext() in D:\k\EntityFramework\src\Ent
ityFramework.Core\Query\EntityQueryExecutor.cs:line 179
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Repro1488.Program.Main(String[] args) in D:\Projects\Repro1488\Repro1488\P
rogram.cs:line 182

**Comment by maumar** [comment]
Most likely same root cause as #1487

**Comment by maumar** [comment]
fixed in 617ad0aaa964949e4a620ef5a73fa4bbe2f12374
